File: app/build_old_105631/tmp/kapt3/stubs/debug/com/example/wardeobe/viewmodel/ProfileViewModel.java
package com.example.wardeobe.viewmodel;

@kotlin.Metadata(mv = {1, 9, 0}, k = 1, xi = 48, d1 = {"\u0000R\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0002\b\u0002\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0002\b\u0003\n\u0002\u0010\u0002\n\u0002\b\u0003\n\u0002\u0010\u000e\n\u0002\b\u0007\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\b\u0007\u0018\u00002\u00020\u0001B\u000f\b\u0007\u0012\u0006\u0010\u0002\u001a\u00020\u0003\u00a2\u0006\u0002\u0010\u0004J\u0006\u0010\u0012\u001a\u00020\u0013J\u0006\u0010\u0014\u001a\u00020\u0013J\u0010\u0010\u0015\u001a\u00020\u00132\u0006\u0010\u0016\u001a\u00020\u0017H\u0002J&\u0010\u0018\u001a\u00020\u00132\u0006\u0010\u0019\u001a\u00020\u00172\u0006\u0010\u001a\u001a\u00020\u00172\u0006\u0010\u001b\u001a\u00020\u0017H\u0082@\u00a2\u0006\u0002\u0010\u001cJ\u0016\u0010\u001d\u001a\u00020\u00132\u0006\u0010\u001e\u001a\u00020\u001f2\u0006\u0010 \u001a\u00020!R\u0014\u0010\u0005\u001a\b\u0012\u0004\u0012\u00020\u00070\u0006X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\b\u001a\u00020\tX\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\n\u001a\u00020\u000bX\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\f\u001a\u00020\rX\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u0002\u001a\u00020\u0003X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u0017\u0010\u000e\u001a\b\u0012\u0004\u0012\u00020\u00070\u000f\u00a2\u0006\b\n\u0000\u001a\u0004\b\u0010\u0010\u0011\u00a8\u0006\""}, d2 = {"Lcom/example/wardeobe/viewmodel/ProfileViewModel;", "Landroidx/lifecycle/ViewModel;", "profileRepository", "Lcom/example/wardeobe/data/ProfileRepository;", "(Lcom/example/wardeobe/data/ProfileRepository;)V", "_uiState", "Lkotlinx/coroutines/flow/MutableStateFlow;", "Lcom/example/wardeobe/viewmodel/ProfileUiState;", "auth", "Lcom/google/firebase/auth/FirebaseAuth;", "cloudinary", "Lcom/cloudinary/Cloudinary;", "firestore", "Lcom/google/firebase/firestore/FirebaseFirestore;", "uiState", "Lkotlinx/coroutines/flow/StateFlow;", "getUiState", "()Lkotlinx/coroutines/flow/StateFlow;", "deleteProfilePicture", "", "fetchUserProfile", "showError", "msg", "", "updateProfileUrlInDatabase", "userId", "url", "publicId", "(Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;Lkotlin/coroutines/Continuation;)Ljava/lang/Object;", "uploadProfilePicture", "context", "Landroid/content/Context;", "uri", "Landroid/net/Uri;", "app_debug"})
@dagger.hilt.android.lifecycle.HiltViewModel()
public final class ProfileViewModel extends androidx.lifecycle.ViewModel {
    @org.jetbrains.annotations.NotNull()
    private final com.example.wardeobe.data.ProfileRepository profileRepository = null;
    @org.jetbrains.annotations.NotNull()
    private final com.google.firebase.auth.FirebaseAuth auth = null;
    @org.jetbrains.annotations.NotNull()
    private final com.google.firebase.firestore.FirebaseFirestore firestore = null;
    @org.jetbrains.annotations.NotNull()
    private final com.cloudinary.Cloudinary cloudinary = null;
    @org.jetbrains.annotations.NotNull()
    private final kotlinx.coroutines.flow.MutableStateFlow<com.example.wardeobe.viewmodel.ProfileUiState> _uiState = null;
    @org.jetbrains.annotations.NotNull()
    private final kotlinx.coroutines.flow.StateFlow<com.example.wardeobe.viewmodel.ProfileUiState> uiState = null;
    
    @javax.inject.Inject()
    public ProfileViewModel(@org.jetbrains.annotations.NotNull()
    com.example.wardeobe.data.ProfileRepository profileRepository) {
        super();
    }
    
    @org.jetbrains.annotations.NotNull()
    public final kotlinx.coroutines.flow.StateFlow<com.example.wardeobe.viewmodel.ProfileUiState> getUiState() {
        return null;
    }
    
    /**
     * Retrieves the current user's profile metadata (VTO URL) from Firestore.
     */
    public final void fetchUserProfile() {
    }
    
    /**
     * Uploads the image to Cloudinary (bypassing Firebase Storage) and updates Firestore.
     */
    public final void uploadProfilePicture(@org.jetbrains.annotations.NotNull()
    android.content.Context context, @org.jetbrains.annotations.NotNull()
    android.net.Uri uri) {
    }
    
    private final java.lang.Object updateProfileUrlInDatabase(java.lang.String userId, java.lang.String url, java.lang.String publicId, kotlin.coroutines.Continuation<? super kotlin.Unit> $completion) {
        return null;
    }
    
    public final void deleteProfilePicture() {
    }
    
    private final void showError(java.lang.String msg) {
    }
}

File: app/src/main/java/com/example/wardeobe/viewmodel/ProfileViewModel.kt
package com.example.wardeobe.viewmodel

import com.example.wardeobe.BuildConfig
import kotlinx.coroutines.CancellationException
import android.content.Context
import android.net.Uri
import android.util.Log
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import javax.inject.Inject
import com.example.wardeobe.model.UserModel
import com.cloudinary.Cloudinary
import com.cloudinary.utils.ObjectUtils
import com.google.firebase.auth.ktx.auth
import com.google.firebase.firestore.ktx.firestore
import com.google.firebase.ktx.Firebase
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.tasks.await
import kotlinx.coroutines.withContext
import java.io.InputStream
import java.io.IOException
import java.util.UUID


import com.google.firebase.functions.FirebaseFunctions

@HiltViewModel
class ProfileViewModel @Inject constructor(
    private val profileRepository: com.example.wardeobe.data.ProfileRepository,
    private val functions: FirebaseFunctions,
    private val auth: com.google.firebase.auth.FirebaseAuth
) : ViewModel() {

    private val firestore = Firebase.firestore

    private val _uiState = MutableStateFlow(ProfileUiState())
    val uiState = _uiState.asStateFlow()

    init {
        fetchUserProfile()
    }

    /**
     * Retrieves the current user's profile metadata (VTO URL) from Firestore.
     */
    fun fetchUserProfile() {
        val userId = auth.currentUser?.uid ?: return

        viewModelScope.launch {
            try {
                // Fetch the existing user profile document
                val document = firestore.collection("users").document(userId).get().await()
                val userModel = document.toObject(UserModel::class.java)

                withContext(Dispatchers.Main) {
                    if (userModel != null) {
                        val url = userModel.profilePictureUrl ?: ""
                        _uiState.value = _uiState.value.copy(
                            profilePictureUrl = url,
                            profilePicturePublicId = userModel.profilePicturePublicId
                        )
                        profileRepository.updateProfilePictureUrl(url)
                    }
                    Log.d("ProfileViewModel", "Profile fetched successfully.")
                }
            } catch (e: CancellationException) { throw e } catch (e: Exception) {
                Log.e("ProfileViewModel", "Error fetching profile: ${e.message}")
            }
        }
    }

    /**
     * Uploads the image via Cloud Function and updates Firestore.
     */
    fun uploadProfilePicture(context: Context, uri: Uri) {
        val userId = auth.currentUser?.uid
        if (userId == null) {
            showError("Please log in to upload a picture.")
            return
        }

        _uiState.value = _uiState.value.copy(loading = true, message = "Uploading image...")

        viewModelScope.launch(Dispatchers.IO) {
            try {
                val compressed = com.example.wardeobe.util.ImageCompressor.compressImage(context, uri)
                if (compressed == null) {
                    throw Exception("Could not compress image.")
                }
                val base64Image = android.util.Base64.encodeToString(compressed, android.util.Base64.NO_WRAP)

                val data = hashMapOf(
                    "imageBase64" to base64Image
                )

                val result = functions
                    .getHttpsCallable("uploadProfilePicture")
                    .call(data)
                    .await()

                val responseData = result.data as? Map<*, *>
                val url = responseData?.get("secure_url") as? String
                val publicId = responseData?.get("public_id") as? String

                if (url == null || publicId == null) {
                    throw Exception("Cloud function upload failed.")
                }

                // 2. Update Firestore UserModel with the new URL
                updateProfileUrlInDatabase(userId, url, publicId)

                withContext(Dispatchers.Main) {
                    _uiState.value = _uiState.value.copy(
                        loading = false,
                        message = "Profile picture saved!",
                        profilePictureUrl = url,
                        profilePicturePublicId = publicId
                    )
                }
            } catch (e: CancellationException) { throw e } catch (e: Exception) {
                Log.e("ProfileViewModel", "Upload failed: ${e.message}", e)
                showError("Upload failed: Check file permissions or try another image.")
            }
        }
    }

    private suspend fun updateProfileUrlInDatabase(userId: String, url: String, publicId: String) {
        val userRef = firestore.collection("users").document(userId)
        val updates = hashMapOf(
            "profilePictureUrl" to url,
            "profilePicturePublicId" to publicId
        )
        userRef.update(updates as Map<String, Any>).await()
    }

    // --- Deletion Logic ---

    fun deleteProfilePicture() {
        val publicId = uiState.value.profilePicturePublicId
        val userId = auth.currentUser?.uid
        if (publicId == null || userId == null) return

        _uiState.value = _uiState.value.copy(loading = true, message = "Deleting picture...")

        viewModelScope.launch(Dispatchers.IO) {
            try {
                val data = hashMapOf("publicId" to publicId)
                functions
                    .getHttpsCallable("deleteProfilePicture")
                    .call(data)
                    .await()

                // Clear the URL in Firestore
                val userRef = firestore.collection("users").document(userId)
                val updates = hashMapOf(
                    "profilePictureUrl" to null,
                    "profilePicturePublicId" to null
                )
                userRef.update(updates as Map<String, Any>).await()

                withContext(Dispatchers.Main) {
                    _uiState.value = _uiState.value.copy(
                        loading = false,
                        message = "Picture removed.",
                        profilePictureUrl = "",
                        profilePicturePublicId = null
                    )
                }
            } catch (e: CancellationException) { throw e } catch (e: Exception) {
                Log.e("ProfileViewModel", "Deletion failed: ${e.message}", e)
                showError("Could not delete picture.")
            }
        }
    }

    private fun showError(msg: String) {
        _uiState.value = _uiState.value.copy(loading = false, message = msg)
    }
}

data class ProfileUiState(
    val loading: Boolean = false,
    val message: String = "",
    val profilePictureUrl: String = "",
    val profilePicturePublicId: String? = null
)